### ISSUE TITLE
which part of the network to train

### ISSUE THREAD

**Issue by hariSky** [facebookresearch/detr#325]
Is there a specific clear way to specify which parts of DETR I want to be trained?
For example in TF there is : 

_The repository lets you choose which part of the network you want to train with which learning rate.
Backbone: The resnet backbone
Transformer: The transforer part
nlayers: The new layers added to finetune the model on your dataset_

Can I the same on pytorch somehow? Especially for the Backbone.

**Comment by alcinos** [comment]
Hi @hariSky 

You'll probably have to modify the code a bit. This is where we define the optimizers for the transformer and the backbone respectively:
https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/main.py#L131-L137
(see that the argument "lr_backbone" only affects the backbone part)

If you want to freeze completely a part of the network you can take inspiration from our freezing mechanism:
https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/models/segmentation.py#L29-L31
For segmentation we freeze the backbone+transformer completely and train only the head. Setting the requires_grad to false on a tensor causes it not to be trained.

I believe I have answered your question, and as such I'm closing this. Feel free to reach out if you have further concerns.
